src/Storages/MergeTree/MergeTreeBackgroundExecutor.h

```diff
@@ -0,0 +1,144 @@
+#pragma once
+
+#include <deque>
+#include <functional>
+#include <atomic>
+#include <mutex>
+#include <future>
+#include <condition_variable>
+#include <set>
+
+#include <boost/circular_buffer.hpp>
+
+#include <common/shared_ptr_helper.h>
+#include <common/logger_useful.h>
+#include <Common/ThreadPool.h>
+#include <Common/Stopwatch.h>
+#include <Storages/MergeTree/IExecutableTask.h>
+
+
+namespace DB
+{
+
+/**
+ *  Executor for a background MergeTree related operations such as merges, mutations, fetches an so on.
+ *  It can execute only successors of ExecutableTask interface.
+ *  Which is a self-written coroutine. It suspends, when returns true from execute() method.
+ *
+ *  Executor consists of ThreadPool to execute pieces of a task (basically calls 'execute' on a task)
+ *  and a scheduler thread, which manages the tasks. Due to bad experience of working with high memory under
+ *  high memory pressure scheduler thread mustn't do any allocations,
+ *  because it will be a fatal error if this thread will die from a random exception.
+ *
+ *  There are two queues of a tasks: pending (main queue for all the tasks) and active (currently executing).
```

Comment:
But it's unclear why we need `pending` tasks? Why we cannot just retry to assign this task if `active` is full? Probably because we can execute task steps one by one?

Reply:
Pending tasks are not a queue of tasks that are failed to be scheduled directly to the pool. Pending tasks are tasks that successfully scheduled to an executor or tasks that have some extra steps to execute. We need this, because the number of tasks will be more than thread to execute. So, when we execute a step of a task, we move this task to pending queue tail and select another task from pending queue head. We also need a queue of active tasks to properly wait for them when a storage shutdowns. 

P.S. I wrote a comment with some `picture`